Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -642,14 +642,17 @@ bool ContainsTyVar(TypeParameter tp, Type ty)
     }
 
     private void CompileDatatypeDestructorsAndAddToInterface(DatatypeDecl dt, ConcreteSyntaxTree wr,
-      ConcreteSyntaxTree interfaceTree, string DtT_TypeArgs) {
+        ConcreteSyntaxTree interfaceTree, string DtT_TypeArgs) {
       foreach (var ctor in dt.Ctors) {
-        foreach (var dtor in ctor.Destructors) {
+        foreach (var (dtor, index) in ctor.Destructors.Select((v, i) => (v, i))) {
           if (dtor.EnclosingCtors[0] == ctor) {
             var arg = dtor.CorrespondingFormals[0];
-            if (!arg.IsGhost && arg.HasName) {
+            if (!arg.IsGhost) {
+              var name = arg.HasName ? arg.CompileName : FormalName(arg, index);
+              var dtor_name = $"dtor_{(arg.HasName ? "" : ctor.CompileName)}{name}"
```

Comment:
Consider renaming to `dtorName` or `destructorName`

Reply:
Consider replacing 
`$"dtor_{(arg.HasName ? "" : ctor.CompileName)}{name}"` 
with
`$"dtor_{ctor.CompileName}{(arg.HasName ? "" : FormalName(arg, index))}"`

I found it hard to read the former since mentally I inline `name`, given that variable name doesn't tell me much, and then it becomes 
`$"dtor_{(arg.HasName ? "" : ctor.CompileName)}{arg.HasName ? arg.CompileName : FormalName(arg, index)}"`

Reply:
Those aren't equivalent.

Reply:
`$"dtor_{(arg.HasName ? "" : ctor.CompileName)}{arg.HasName ? arg.CompileName : FormalName(arg, index)}"`

if `arg.HasName` is true that becomes:
`$"dtor_{""}{arg.CompileName}"` => `$"dtor_{arg.CompileName}"` 

and otherwise:
`$"dtor_{ctor.CompileName}{FormalName(arg, index)}"`

So.. you could replace that with
`$"dtor_{ctor.CompileName}{arg.HasName ? "" : FormalName(arg, index)}"`

right?

Reply:
Otherwise, one uses `arg` the other `ctor`.

Reply:
I see, thanks.